dnf/transaction_sr.py

```diff
@@ -0,0 +1,169 @@
+            _('Incompatible minor version {minor}, '
+                'supported minor version is {minor_supp} or greater.'
+            ).format(minor=minor, minor_supp=VERSION_MINOR)
+        )
+
+
+def store_transaction(transaction, filename="transaction.json"):
+    data = {
+        "version": VERSION,
+    }
+    rpms = {}
+    groups = {}
+    environments = {}
+
+    for tsi in transaction.packages():
+        if tsi.is_package():
+            na = "%s.%s" % (tsi.name, tsi.arch)
+
+            data_p = rpms.setdefault(na, {})
+
+            # for installonly packages the package data needs to be a list of actions
+            if isinstance(data_p, list):
+                data_p.append({})
+                data_p = data_p[-1]
+
+            if tsi._item.isForwardAction():
+                if "to" in data_p or "action" in data_p and data_p["action"] == "Removed":
+                    # installonly pkg, convert to a list of actions
+                    rpms[na] = [data_p, {}]
+                    data_p = rpms[na][-1]
+
+                data_p["to"] = tsi.evr
+                data_p["action"] = tsi.action_name
+            elif tsi._item.isBackwardAction():
+                if "from" in data_p or "action" in data_p and data_p["action"] == "Install":
+                    # installonly pkg, convert to a list of actions
+                    rpms[na] = [data_p, {}]
+                    data_p = rpms[na][-1]
+
+                data_p["from"] = tsi.evr
+                # make sure not to overwrite the forward action if it was written already
+                if "action" not in data_p:
+                    data_p["action"] = tsi.action_name
+            else:
+                data_p["weird_reason"] = True
+
+            data_p["reason"] = libdnf.transaction.TransactionItemReasonToString(tsi.reason)
+
+        elif tsi.is_group():
+            group = tsi.get_group()
+
+            data_p = groups.setdefault(group.getGroupId(), {})
+            data_p["action"] = tsi.action_name
+
+            if tsi.action == libdnf.transaction.TransactionItemAction_INSTALL:
+                data_p["package_types"] = \
+                    libdnf.transaction.compsPackageTypeToString(group.getPackageTypes())
+
+        elif tsi.is_environment():
+            env = tsi.get_environment()
+
+            data_p = environments.setdefault(env.getEnvironmentId(), {})
+            data_p["action"] = tsi.action_name
+
+            if tsi.action == libdnf.transaction.TransactionItemAction_INSTALL:
+                data_p["package_types"] = \
+                    libdnf.transaction.compsPackageTypeToString(env.getPackageTypes())
+
+    if rpms:
+        data["rpms"] = rpms
+
+    if groups:
+        data["groups"] = groups
+
+    if environments:
+        data["environments"] = environments
+
+    with open(filename, "w") as f:
+        json.dump(data, f, indent=4, sort_keys=True)
+        f.write("\n")
+
+    print(_("Transaction saved to {}.").format(filename))
```

Comment:
What about to provide a full path?

Reply:
I would prefer to provide full path especial when in future there could be an option that will define file path.

Reply:
@lukash Please could you provide a full path in output? yum also provided the full path. See bellow. In case that full path is not provided it will require that user will have to search for it, look ad man pages or so on. The information should be always in form that does not require any research from end user side.
```
Your transaction was saved, rerun it with:
 yum load-transaction /tmp/yum_save_tx.2020-04-01.12-36.CT5VMR.yumtx
```

Reply:
To be honest I just don't see how the user could not find the file. The file is saved in the current directory and the filename is printed out to the output, so it's very clear where the file is.

The yum situation just doesn't apply, it isn't saving the file to the current working directory, so it certainly needs to print the full path.

For the case of running the command from a script, the script will just not parse the output to find the file. It will expect to find the file at the default location (`transaction.json` in the current directory) or, more likely, use the `-o` switch (which I should just add straight away) to define the file location.

TL;DR I find it completely unambiguous to print the relative path and printing an absolute path will just add more characters to the output for the user to parse... Unless I'm missing a case where the full path is actually useful.

Reply:
I believe that `store_transaction(transaction, filename="transaction.json")` will be also used by system-upgrade. Therefore there should be not `print(_("Transaction saved to {}.").format(filename))` because it will be confusing, because it will report internal information that should be hidden for `system-upgrade download`.

Reply:
@lukash After some discussions I realized other thing, that `dnf history store` overwrites content of "transaction.json" without any note. Now I understand why yum creates transaction file in `/tmp`. What about to use the same approach and create unique tmp file/directory to store the transaction?

Reply:
I've got this sorted out locally already. I've moved the code around, so the file saving is in the history command, not in the API function, added the `-o` option to specify the output file and I've added an overwrite check with a a user prompt in case the file already exists. This is the standard approach... (except maybe typically CLI tools would print to stdout if no `-o` is given, we probably don't want to do that)